File: Python/binaryTree/binarySearchTree/95.unique-binary-search-trees-ii.py
#
# @lc app=leetcode id=95 lang=python3
#
# [95] Unique Binary Search Trees II
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    uniqueDict = {(0, 0):[None]}
    def generateTreesRecur(self, low: int, high: int) -> List[Optional[TreeNode]]:
        if (low, high) in self.uniqueDict: 
            return self.uniqueDict[(low, high)]
        else: 
            for rootVal in range(low, high+1):
                leftTreeRange = (low, rootVal-1)
                rightTreeRange = (rootVal+1, high)
                if rootVal == low:
                    leftTreeRange = (0, 0)
                if rootVal == high:
                    rightTreeRange = (0, 0)

                leftTreesList = self.generateTreesRecur(leftTreeRange[0], leftTreeRange[1])
                rightTreesList = self.generateTreesRecur(rightTreeRange[0], rightTreeRange[1])


                for treeL in leftTreesList:
                    for treeR in rightTreesList:
                        root = TreeNode(rootVal)
                        root.left = treeL
                        root.right = treeR
                        if ((low, high) not in self.uniqueDict):
                            self.uniqueDict[(low, high)] = [root]
                        else:
                            self.uniqueDict[(low, high)].append(root)

            return self.uniqueDict[(low, high)]


    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        
        return self.generateTreesRecur(1, n)



        
# @lc code=end

# 如果要优化的话可以用二叉树sequence(codec)算法把每一个结构unique的树编码

File: LeetCode_Go/BinaryTree/95.unique-binary-search-trees-ii.go
/*
 * @lc app=leetcode id=95 lang=golang
 *
 * [95] Unique Binary Search Trees II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// Given an integer n, return all the structurally unique BST's, which has exactly n nodes of unique values from 1 to n in an array.

// recursive function build all possible BSTs from continuous valued node from `start` to `end`. For any possible `root`, we build all possible BSTs for the left subtree and the right subtree, and then do a combination to connect the root and the left/right subtrees.
// rewrite to DP: use a read-only DP table to store the result (`templates`), and use a function to copy trees from the DP table.

// O(n^2)
func generateTrees(n int) []*TreeNode {
	subStructures := make(map[[2]int][]*TreeNode)
	return generateTreesHelper(1, n, subStructures)
}

func generateTreesHelper(start, end int, subStructures map[[2]int][]*TreeNode) []*TreeNode {
	// base case
	if start > end {
		return []*TreeNode{nil}
	}

	var result []*TreeNode

	// check DP table
	key := [2]int{start, end}
	if value, exists := subStructures[key]; exists {
		for _, root := range value {
			result = append(result, cloneTree(root))
		}
		return result
	}

	// iterate subtrees with mid as root
	for mid := start; mid <= end; mid++ {
		leftStructures := generateTreesHelper(start, mid-1, subStructures)
		rightStructures := generateTreesHelper(mid+1, end, subStructures)

		// combination, connecting sub-structures
		for i := 0; i < len(leftStructures); i++ {
			for j := 0; j < len(rightStructures); j++ {
				root := &TreeNode{Val: mid}
				root.Left = leftStructures[i]
				root.Right = rightStructures[j]
				result = append(result, root)
			}
		}
	}
	subStructures[key] = result
	return result

}

func cloneTree(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	newRoot := &TreeNode{Val: root.Val}
	newRoot.Left = cloneTree(root.Left)
	newRoot.Right = cloneTree(root.Right)
	return newRoot
}

// @lc code=end

